psi4/driver/broker.py

```diff
@@ -0,0 +1,159 @@
+#
+# @BEGIN LICENSE
+#
+# Psi4: an open-source quantum chemistry software package
+#
+# Copyright (c) 2013-2018 The Psi4 Developers.
+#
+# The copyrights for code used from other parties are included in
+# the corresponding files.
+#
+# This file is part of Psi4.
+#
+# Psi4 is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, version 3.
+#
+# Psi4 is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with Psi4; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# @END LICENSE
+#
+
+from __future__ import print_function
+
+import string
+import sys
+import time
+import numpy as np
+from numpy.testing import assert_array_almost_equal as assert_equal
+
+import psi4
+try:
+    from ipi.interfaces.sockets import Client
+    ipi_available = True
+except ImportError:
+    ipi_available = False
+    # Define Client to enable testing of the Broker in the unittests
+    class Client(object): pass
+
+
+class Broker(Client):
+    def __init__(self, options, genbas=None, serverdata=False):
+        self.serverdata = serverdata
+        if not ipi_available:
+            print("i-pi is not available for import:",
+                  "The broker infrastructure will not be available!")
+            super(Broker, self).__init__()
+        elif serverdata:
+            mode, address, port = serverdata.split(":")
+            mode = string.lower(mode)
+            super(Broker, self).__init__(address=address, port=port,
+                                      mode=mode)
+        else:
+            super(Broker, self).__init__(_socket=False)
+        self.options = options
+
+        print("PSI4 options:")
+        for item, value in self.options.items():
+            print(item, value)
+            if item not in ["LOT", "multiplicity", "charge"]:
+                psi4.core.set_global_option(item, value)
+        psi4.core.IO.set_default_namespace("xwrapper")
+
+        self.timing = {}
+
+    def calculate_force(self, pos):
+        """Fetch force, energy of PSI.
+        """
+        if len(pos.shape) == 1:
+            pos = pos.reshape((-1, 3))
+        self.frc, self.pot = self.callback(pos)
+        return self.frc, self.pot
+
+    def get_molecule(self, pos):
```

Comment:
If you pass in a Psi4 molecule perhaps a better way to instantiate this will be `molclone = self.mol.clone(); mol clone.set_geometry(psi4.core.Matrix.from_array(pos))`. In this case all of the multiplicity, symmetry, etc will be carried over automatically.

Reply:
Hmm, maybe `get_molecule` is a bit misleading as a function name:
I would like to have the final string representation in cartesian coordinates with c1 symmetry that can be used to calculate the force. The reason is, that the molecule will mostly have no symmetry anyway due to the dynamics and this will simplify other algorithms to speed up the force calculation by reusing previous orbitals without changing symmetry/charge/multiplicity.

Would you prefer to name this function like `get_geometry` instead?

Reply:
do you need the string repr for any other reason than to pass into `psi4.geometry(` to create a new active molecule?

if you need the string, an easier way is `molinstance.to_string(dtype='xyz', units='bohr')`. if you're after an actual `Molecule` with the new geometry, @dgasmith's suggestion is the way to go. Has the parent molecule already had c1/nocom/noreorient set?